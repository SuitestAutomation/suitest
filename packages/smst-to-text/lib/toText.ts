/// <reference path="../../smst/types/unistTestLine.d.ts" />

import {assertUnknownSectionNode} from '@suitest/smst/commonjs/jsxFactory';

type RenderTextFunc = (node: ExtendedInlineNodes) => string;

type ExtendedInlineNodes = InlineTextNode | {
	type: TestLineResultStatus | SingleEntryStatus,
	value: string,
};

const nl = '\n';
const tab = '  ';

const format = {
	cancel: '\u001b[0m',
	subject: '\u001b[32m',	// green
	code: '\u001b[36m',	// cyan
	input: '\u001b[4m',	// underscore
	success: '\u001b[32m',	// green
	fail: '\u001b[31m',	// red
	fatal: '\u001b[31m',	// red
	warning: '\u001b[33m',	// yellow
	exit: '\u001b[34m',	// blue
};

const formatString = (text: string, type: string): string => {
	if (type in format) {
		return format[type as keyof typeof format] + text + format.cancel;
	}

	return text;
};

/**
 * Render a single text node as plain text
 */
const renderPlainTextNode: RenderTextFunc = (node: ExtendedInlineNodes): string => node.value;

/**
 * Render a single text node with ANSI styling
 */
const renderFormattedTextNode: RenderTextFunc = (node: ExtendedInlineNodes): string =>
	formatString(node.value ?? '', node.type);

const renderStatus = (type: TestLineResultStatus | SingleEntryStatus): ExtendedInlineNodes => {
	let value = '';

	switch (type) {
		case 'success':
			value = '✔ ';
			break;
		case 'fail':
		case 'fatal':
			value = '✖ ';
			break;
		case 'warning':
			value = '‼ ';
			break;
		case 'excluded':
		case 'exit':
			value = '» ';
			break;
	}

	return {type, value};
};

/**
 * Split a single AST node into several based on provided length of the first node
 */
const splitNode = (textNode: ExtendedInlineNodes, length: number): [ExtendedInlineNodes, ExtendedInlineNodes] => ([
	{
		...textNode,
		value: textNode.value.slice(0, length),
	},
	{
		...textNode,
		value: textNode.value.slice(length),
	},
]);

/**
 * Wrapped textual nodes to fit max length
 * @TODO consider wrapping by word and not by any character
 */
const wrapTextNodes = (
	textNodes: ExtendedInlineNodes[],
	renderTextNode: RenderTextFunc,
	maxLineLength = 60
): [number, string[]] => {
	// Keep function immutable
	textNodes = [...textNodes];

	const output: string[] = [''];
	let maxActualLength = 0;
	let currentLineLength = 0;

	let firstNode = textNodes.shift();
	while (firstNode) {
		if (firstNode.value.length < maxLineLength - currentLineLength) {
			// The whole textNode can fit into the line
			currentLineLength += firstNode.value.length;
			output[output.length - 1] += renderTextNode(firstNode);
			maxActualLength = currentLineLength > maxActualLength ? currentLineLength : maxActualLength;
		} else {
			// Need to split text node
			const splitNodes = splitNode(firstNode, maxLineLength - currentLineLength);
			currentLineLength = 0;
			output[output.length - 1] += renderTextNode(splitNodes[0]);
			output.push('');
			textNodes.splice(0, 1, splitNodes[1]);
			maxActualLength = maxLineLength;
		}

		firstNode = textNodes.shift();
	}

	return [maxActualLength, output];
};

const renderProps = (node: PropertiesNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	const columnsLength: [number, number, number] = [0, 0, 0];
	// Level 1 = table fragment (several rows)
	// Level 2 = single table row
	// Level 3 = lines in a single cell
	const table: string[][][] = [];

	// Go through each prop and instead produce a bunch of rows for table rendering
	for (const prop of node.children) {
		let currentRow: string[][] = [];
		table.push(currentRow);

		// Process first cell - property name
		const propNameColumn: ExtendedInlineNodes[] = [...prop.name];

		if (prop.status) {
			propNameColumn.splice(0, 0, renderStatus(prop.status));
		}

		const [nameCellLength, nameColumnContent] = wrapTextNodes(propNameColumn, renderTextNode);
		currentRow.push(nameColumnContent);

		if (nameCellLength > columnsLength[0]) {
			columnsLength[0] = nameCellLength;
		}

		// Process second column - comparator
		const comparatorCellLength = prop.comparator?.length ?? 0;
		currentRow.push([prop.comparator ?? '']);

		if (comparatorCellLength > columnsLength[1]) {
			columnsLength[1] = comparatorCellLength;
		}

		// Process third column - expected value
		if (prop.contentType === 'block') {
			// Render new line with a single cell - code cell
			currentRow = [];
			table.push(currentRow);

			// Currently there is only one type of block content in SMST - code block
			// If that ever changes, this has to be updated
			currentRow.push(prop.expectedValue.value.split(nl).map(line => tab + '> ' + line));
		} else {
			// Render non-code block value
			const [valueCellLength, valueColumnContent] = wrapTextNodes(prop.expectedValue, renderTextNode);
			currentRow.push(valueColumnContent);

			if (valueCellLength > columnsLength[2]) {
				columnsLength[2] = valueCellLength;
			}
		}

		// Add actual value to the mix as a separate row
		if (prop.contentType === 'inline' && prop.actualValue) {
			// Add new row
			currentRow = [];
			table.push(currentRow);

			// property name cell is empty, comparator cell is always an arrow
			currentRow.push([], ['→']);

			// Do not forget to increase total column width, in case previous row did not
			// have a comparator somehow
			if (columnsLength[1] === 0) {
				columnsLength[1] = 1;
			}

			// Render the actual value
			const [valueCellLength, valueColumnContent] = wrapTextNodes([{
				type: 'text' as const,
				value: String(prop.actualValue),
			}], renderTextNode);
			currentRow.push(valueColumnContent);

			if (valueCellLength > columnsLength[2]) {
				columnsLength[2] = valueCellLength;
			}
		}
	}

	// Now convert the table into a list of text lines
	const lines: string[] = [];

	for (const row of table) {
		while (row.some(cell => cell.length)) {
			// While there is a non-empty line in at least one cell
			lines.push(prefix + row
				.map((cell, columnIndex) => {
					const cellLine = cell.shift() ?? '';

					return cellLine.padEnd(columnsLength[columnIndex]);
				})
				.join(' ')
			);
		}
	}

	return lines.join(nl);
};

const renderNode = (node: SingleNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	switch (node.type) {
		case 'text':
		case 'code':
		case 'subject':
		case 'input':
			return renderTextNode(node);
		case 'props':
			return renderProps(node, renderTextNode, prefix);
		case 'prop':
			throw new Error('Prop node can only be rendered as part of Props');
		case 'code-block':
			throw new Error('CodeBlock node can only be rendered as part of Props');
		case 'test-line':
			return renderTestLineOrCondition(node, renderTextNode);
		case 'condition':
			return nl + prefix + renderTestLineOrCondition(node, renderTextNode, prefix + tab);
		case 'test-line-result':
			return renderTestLineResult(node, renderTextNode, prefix);
		default:
			/* istanbul ignore next */
			return assertUnknownSectionNode(node);
	}
};

const renderTestLineOrCondition = (
	node: TestLineNode | ConditionNode,
	renderTextNode: RenderTextFunc,
	prefix = ''
): string => {
	const status = node.status ? renderTextNode(renderStatus(node.status)) : '';
	const title = node.title.map(renderTextNode).join('');
	const body = node.children.map(child => renderNode(child, renderTextNode, prefix + tab)).join('');

	return status + title + nl + body;
};

const renderTestLineResult = (node: TestLineResultNode, renderTextNode: RenderTextFunc, prefix = ''): string => {
	const message = node.message
		? renderTextNode({type: node.status, value: node.message.map(renderTextNode).join('')})
		: '';
	const body = renderTestLineOrCondition(node.children[0], renderTextNode, prefix + tab);

	return [body, message].filter(Boolean).join(nl);
};

export const toText = (node: Node, format = true): string => {
	const renderTextNode = format ? renderFormattedTextNode : renderPlainTextNode;

	if (!Array.isArray(node)) {
		node = [node];
	}

	return node.map(n => renderNode(n, renderTextNode)).join('');
};
