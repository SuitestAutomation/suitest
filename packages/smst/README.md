# SMST

**S**uitest **m**essage **s**yntax **t**ree.

Suitest flavour of the [unist][ext-unist] domain-specific syntax tree, designed
to express test line definitions, results and other user-facing messages. SMST can
be used to render messages in plain text, text with ANSI styling and HTML.

## Nodes

### `Text`

Inline plain text.

```idl
interface Text {
    type: "text"
    value: string
}
```

### `Subject`

Inline text, that represents test line subject name

```idl
interface Subject <: Text {
    type: "subject"
}
```

See [Text][dfn-text].

### `Input`

Inline text, that represents user-defined values in test lines, e.g. expectations.

```idl
interface Input <: Text {
    type: "input"
}
```

See [Text][dfn-text].

### `Code`

Inline code.

```idl
interface Code <: Text {
    type: "code"
}
```

See [Text][dfn-text].

### `Paragraph`

A collection of simple text nodes.

```idl
interface Paragraph {
    type: "paragraph"
    children: InlineText[]
}
```

See [InlineText][dfn-inline-text]

### `CodeBlock`

A block of code.

```idl
interface CodeBlock {
    type: "code-block"
    language: "javascript" | "brightscript"
    value: string
}
```

See [Paragraph][dfn-paragraph].

### `Property`

A single prop in the [properties table][dfn-properties].

```idl
interface Property {
    type: "prop"
    prop: Paragraph
    comparator: string
    expectedValue: Paragraph
    actualValue: [Paragraph | CodeBlock]
    status: ["success" | "failure"]
}
``` 

See [Paragraph][dfn-paragraph] and [CodeBlock][dfn-code-block].

### `Properties`

A simple table, where each row represents a property  

```idl
interface Properties {
    type: "props"
    children: Property[]
}
```

See [Property][dfn-property].

### `Condition`

A test line condition. Could be used in assertions or as a part of conditional line.
For example, `element [my element] exists` is a condition and can be used in such lines:

* `Assert element [my element] exists timeout 2s`
* `Press button OK only if element [my element] exists`

Condition consists of title and optional [Properties][dfn-properties] table. Could
also include result status when rendering test line result. 

```idl
interface Condition {
    type: "condition"
    title: Paragraph
    children: [Properties]
    status: ["success" | "fail"]
}
```

See [Paragraph][dfn-paragraph] and [Properties][dfn-properties].

### `TestLine`

Represent a single Suitest test line.

```idl
interface TestLine {
    type: "test-line"
    title: Paragraph
    children: [(Properties | Condition)[]]
    status: ["success" | "fatal" | "fail" | "warning" | "exit" | "excluded"]
}
```

See [Paragraph][dfn-paragraph], [Properties][dfn-properties] and [Condition][dfn-condition]

### `TestLineResult`

Test line result is a wrapper around [TestLine][dfn-test-line], that also includes execution results
for that line.

```idl
interface TestLineResult {
    type: "test-line-result"
    level: "success" | "fatal" | "fail" | "warning" | "exit" | "excluded"
    children: TestLine
    message: [Paragraph],
    screenhost: [string],
}
```

See [Paragraph][dfn-paragraph] and [TestLine][dfn-test-line]

## Enumeration

### `InlineText`

Any inline text node.

```idl
type InlineText = Text | Subject | Input | Code
```

[ext-unist]: https://github.com/syntax-tree/unist
[dfn-text]: #text
[dfn-paragraph]: #paragraph
[dfn-code-block]: #codeblock
[dfn-inline-text]: #inlinetext
[dfn-test-line]: #testline
[dfn-property]: #property
[dfn-properties]: #properties
