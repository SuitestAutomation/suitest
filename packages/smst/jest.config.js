module.exports = {
    "preset": "ts-jest",
    "transform": {
        "^.+\\.(j|t)sx?$": "ts-jest"
    },
    "testMatch": [
        "<rootDir>/lib/__tests__/*.test.ts",
		"<rootDir>/lib/__tests__/*.test.tsx"
    ],
    "collectCoverageFrom": [
        "<rootDir>/lib/**/*.{ts,tsx}",
		"!<rootDir>/lib/__tests__/**/*"
    ],
    "coverageThreshold": {
        "global": {
            "branches": 90,
            "functions": 95,
            "lines": 95,
            "statements": 95
        }
    },
	"moduleFileExtensions": [
		"js", "json", "jsx", "ts", "d.ts", "tsx", "node",
	],
	"globals": {
		"ts-jest": {
			"diagnostics": {
				"warnOnly": true
			},
		},
	},
};
