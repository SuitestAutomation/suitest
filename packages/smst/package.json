{
  "name": "@suitest/smst",
  "version": "3.0.2",
  "description": "UNIST definition and JSX factory for Suitest test lines rendering",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/SuitestAutomation/suitest.git",
    "directory": "packages/smst"
  },
  "kywords": [
    "suitest",
    "unist"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/SuitestAutomation/suitest/issues"
  },
  "homepage": "https://suite.st/",
  "main": "lib/jsxFactory.js",
  "files": [
    "umd/",
    "commonjs/",
    "types/"
  ],
  "scripts": {
    "build": "npm run build:umd && npm run build:commonjs",
    "build:umd": "rimraf umd && rollup -c --environment FORMAT:umd",
    "build:commonjs": "rimraf commonjs && rollup -c --environment FORMAT:commonjs",
    "test": "jest",
    "coverage": "jest --coverage",
    "prepublishOnly": "npm run coverage && npm run build"
  },
  "dependencies": {
    "unist-builder": "^2.0.3"
  },
  "devDependencies": {
    "rollup": "^1.21.4",
    "rimraf": "3.0.0"
  }
}
