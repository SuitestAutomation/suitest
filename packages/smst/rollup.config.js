import typescript from 'rollup-plugin-typescript2';
import { uglify } from "rollup-plugin-uglify";

export default {
	input: './lib/jsxFactory.ts',
	external: [
		'unist-builder',
	],
	plugins: [
		uglify(),
		typescript({tsconfigOverride: {compilerOptions: {module: 'ES2015'}}}),
	],
	output: {
		file: `./${process.env.FORMAT}/jsxFactory.js`,
		name: 'translate',
		format: process.env.FORMAT,
		compact: process.env.FORMAT === 'umd',
		globals: {
			'unist-builder': 'ub',
		},
	},
};
