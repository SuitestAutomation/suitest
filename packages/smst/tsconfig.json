{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "lib",
    "outDir": "commonjs",
    "moduleResolution": "node"
  },
  "include": [
    "lib",
    "types"
  ],
  "exclude": [
    "node_modules",
    "coverage",
    "umd",
    "commonjs",
    "lib/__tests__"
  ]
}
