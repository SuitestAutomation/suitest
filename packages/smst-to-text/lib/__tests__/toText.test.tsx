import {jsx} from '@suitest/smst';
import {toText} from '../toText';

describe('AST renderers', () => {
	const plainText = <text>TEXT</text>;
	const subjectText = <subject>SUBJECT</subject>;
	const inputText = <input>INPUT</input>;
	const codeText = <code>CODE</code>;
	const mixedText = <fragment>TEXT <subject>SUBJ</subject> TEXT</fragment>;
	const emptyText = <text>{''}</text>;

	const simpleCodeBlock = <code-block>someJS();</code-block>;
	const longCodeBlock = <code-block>
		someJS(); someJS(); someJS(); someJS();{'\n'}someJS(); someJS(); someJS();{'\n'}someJS(); someJS(); someJS();
	</code-block>;

	const simpleProps = <props>
		<prop
			name={<text>prop name</text>}
			expectedValue={<text>expected value</text>}
			comparator="="
		/>
		<prop
			name={<text>prop name</text>}
			expectedValue={simpleCodeBlock}
			comparator=""
		/>
	</props>;
	const longProps = <props>
		<prop
			name={<text>prop name prop name prop name prop name prop name</text>}
			expectedValue={<text>expected value expected value expected value expected value expected value</text>}
			comparator="~"
			actualValue="expected value expected value expected value expected value expected value"
			status="fail"
		/>
		<prop
			name={<text>JavaScript expression</text>}
			expectedValue={longCodeBlock}
			comparator=""
			status="success"
		/>
	</props>;

	const simpleCondition = <condition
		title={<fragment>condition: element <subject>My element</subject> exists</fragment>}
	/>;
	const longCondition = <condition
		title={<fragment>condition: element <subject>My element</subject> exists</fragment>}
		status="fail"
	>{longProps}</condition>;
	const simpleLine = <test-line
		title={<fragment>Assert element <subject>My element</subject> is visible</fragment>}
	/>;
	const longLine = <test-line
		title={<fragment>Press <input>OK</input>, <input>LEFT</input> only if condition is met</fragment>}
	>
		{simpleCondition}
	</test-line>;

	const failResult = (screenshot?: string): JSX.Element => <test-line-result
		status="fail"
		message={<text>Condition was not met</text>}
		screenshot={screenshot}
	>
		<test-line
			title={<fragment>Run test <subject>My test</subject> until condition is met max 5x every 5s</fragment>}
			status="fail"
		>{longCondition}</test-line>
	</test-line-result>;
	const warningResult = (screenshot?: string): JSX.Element => <test-line-result
		status="warning"
		message={<text>Some warning message</text>}
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="warning"
		/>
	</test-line-result>;
	const exitResult = (screenshot?: string): JSX.Element => <test-line-result
		status="exit"
		message={<text>Condition was met</text>}
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="exit"
		/>
	</test-line-result>;
	const excludedResult = (screenshot?: string): JSX.Element => <test-line-result
		status="excluded"
		message={<text>Line was not executed</text>}
		screenshot={screenshot}
	>
		<test-line
			title={<fragment>Press <input>OK</input> only if condition is met</fragment>}
			status="excluded"
		>{simpleCondition}</test-line>
	</test-line-result>;
	const fatalResult = (screenshot?: string): JSX.Element => <test-line-result
		status="fatal"
		message={<text>Some generic error message</text>}
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="fatal"
		/>
	</test-line-result>;
	const successResultWithScreenshot = (screenshot?: string): JSX.Element => <test-line-result
		status="success"
		screenshot={screenshot}
	>
		<test-line
			title={<text>Assert application has exited</text>}
			status="success"
		/>
	</test-line-result>;

	function runTests(formatted = false): void {
		it('should handle code blocks', () => {
			expect(toText(simpleCodeBlock, formatted)).toMatchSnapshot();
			expect(toText(longCodeBlock, formatted)).toMatchSnapshot();
		});

		it('should handle props', () => {
			expect(toText(simpleProps, formatted)).toMatchSnapshot();
			expect(toText(longProps, formatted)).toMatchSnapshot();
			expect(toText(<props>
				<prop
					name={<text>test</text>}
					expectedValue={<text>short</text>}
					actualValue="long long long"
				/>
			</props>, formatted)).toMatchSnapshot();
			expect(toText(<props>
				<prop
					name={<text>Empty string</text>}
					expectedValue={<text>{''}</text>}
					actualValue={''}
				/>
				<prop
					name={<text>Empty number</text>}
					expectedValue={<text>0</text>}
					actualValue={NaN}
				/>
				<prop
					name={<text>Empty code block</text>}
					expectedValue={<fragment><code>{'looooooooooooooo oooooooo ooooooo ooooo ooooo oooooooooo oooo oooo ng'}</code> (and some next text block)</fragment>}
				/>
			</props>)).toMatchSnapshot();
		});

		it('should throw if trying to render a single prop', () => {
			expect(() => toText(<prop name={<text>prop</text>} expectedValue={<text>test</text>} />)).toThrow();
		});

		it('should render condition', () => {
			expect(toText(simpleCondition, formatted)).toMatchSnapshot();
			expect(toText(longCondition, formatted)).toMatchSnapshot();
		});

		it('should render test line', () => {
			expect(toText(simpleLine, formatted)).toMatchSnapshot();
			expect(toText(longLine, formatted)).toMatchSnapshot();
		});

		it('should render test line results', () => {
			expect(toText(failResult(), formatted)).toMatchSnapshot();
			expect(toText(warningResult(), formatted)).toMatchSnapshot();
			expect(toText(exitResult(), formatted)).toMatchSnapshot();
			expect(toText(excludedResult(), formatted)).toMatchSnapshot();
			expect(toText(fatalResult(), formatted)).toMatchSnapshot();
			expect(toText(successResultWithScreenshot(), formatted)).toMatchSnapshot();
		});

		it('should render test line results with screenshots', () => {
			const screenshot = 'path/to/screenshot.png';
			expect(toText(failResult(screenshot), formatted)).toMatchSnapshot();
			expect(toText(warningResult(screenshot), formatted)).toMatchSnapshot();
			expect(toText(exitResult(screenshot), formatted)).toMatchSnapshot();
			expect(toText(excludedResult(screenshot), formatted)).toMatchSnapshot();
			expect(toText(fatalResult(screenshot), formatted)).toMatchSnapshot();
			expect(toText(successResultWithScreenshot(screenshot), formatted)).toMatchSnapshot();
		});
	}

	describe('plain text renderer', () => {
		it('should handle textual nodes', () => {
			expect(toText(plainText, false)).toEqual('TEXT');
			expect(toText(subjectText, false)).toEqual('SUBJECT');
			expect(toText(inputText, false)).toEqual('INPUT');
			expect(toText(codeText, false)).toEqual('CODE');
			expect(toText(mixedText, false)).toEqual('TEXT SUBJ TEXT');
			expect(toText(emptyText, false)).toEqual('');
		});

		runTests(false);
	});

	describe('formatted text renderer', () => {
		it('should handle textual nodes', () => {
			expect(toText(plainText, true)).toEqual('TEXT');
			expect(toText(subjectText, true)).toEqual('\u001b[32mSUBJECT\u001b[0m');
			expect(toText(inputText, true)).toEqual('\u001b[4mINPUT\u001b[0m');
			expect(toText(codeText, true)).toEqual('\u001b[36mCODE\u001b[0m');
			expect(toText(mixedText, true)).toEqual('TEXT \u001b[32mSUBJ\u001b[0m TEXT');
			expect(toText(emptyText, true)).toEqual('');
		});

		runTests(true);
	});

	describe('wrapped title and error messages', () => {
		const renderLongTexts = (format: boolean): void => {
			expect(toText(<test-line-result
				status="fail"
				message={<text>Subject does not exist a very loooo oooooo ooooo ooooooo oooooo oooooo oooong error
					message, possibly including "description" part for notStarterReason</text>}
			>
				<test-line
					title={<fragment>Run test
						<subject>My loooooooooo ooooo ooo oooooo ooooooo ooooooo ooooooo ooooo ooooo ng!
							test</subject> until condition is met max 5x every 5s</fragment>}
					status="fail"
				>{longCondition}</test-line>
			</test-line-result>, format)).toMatchSnapshot();
			expect(toText(<test-line-result
				status="fail"
				message={<text>Subject does not exist a very not so loooooooong error message,
					possibly including "description" part for notStarterReason</text>}
			>
				<test-line
					title={<fragment>Run test <subject>My not so loooooooooong! test</subject> until condition
						is met max 5x every 5s</fragment>}
					status="fail"
				>{longCondition}</test-line>
			</test-line-result>, format)).toMatchSnapshot();
		};

		renderLongTexts(true);
		renderLongTexts(false);
	});

	it('render link', () => {
		expect(toText(<link href="http://some.url">Some URL</link>)).toEqual('Some URL (http://some.url)');
		expect(toText(<link href="http://some.url">http://some.url</link>)).toEqual('http://some.url');
		expect(toText(<link href="http://some.url"/>)).toEqual('http://some.url');
	});
});
