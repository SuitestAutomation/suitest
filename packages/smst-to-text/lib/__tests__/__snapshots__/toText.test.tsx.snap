// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST renderers formatted text renderer should handle tables 1`] = `
"prop name = expected value
prop name  
  > someJS();"
`;

exports[`AST renderers formatted text renderer should handle tables 2`] = `
"[31mâœ– [0mprop name prop name prop name prop name prop name ~ expected value expected value expected value expected value 
                                                      expected value                                              
                                                    â†’ expected value expected value expected value expected value 
                                                      expected value                                              
[31mâœ– [0mJavaScript expression                     
  > someJS(); someJS(); someJS(); someJS();        
  > someJS(); someJS(); someJS();                  
  > someJS(); someJS(); someJS();                  "
`;

exports[`AST renderers formatted text renderer should render condition 1`] = `
"
condition: element [32mMy element[0m exists
"
`;

exports[`AST renderers formatted text renderer should render condition 2`] = `
"
condition: element [32mMy element[0m exists
    [31mâœ– [0mprop name prop name prop name prop name prop name ~ expected value expected value expected value expected value 
                                                          expected value                                              
                                                        â†’ expected value expected value expected value expected value 
                                                          expected value                                              
    [31mâœ– [0mJavaScript expression                     
      > someJS(); someJS(); someJS(); someJS();        
      > someJS(); someJS(); someJS();                  
      > someJS(); someJS(); someJS();                  "
`;

exports[`AST renderers formatted text renderer should render test line 1`] = `
"Assert element [32mMy element[0m is visible
"
`;

exports[`AST renderers formatted text renderer should render test line 2`] = `
"Assert element [32mMy element[0m

  condition: element [32mMy element[0m exists
"
`;

exports[`AST renderers formatted text renderer should render test line results 1`] = `
"[31mâœ– [0mAssert application has exited

[31mfail: Condition was not met[0m"
`;

exports[`AST renderers formatted text renderer should render test line results 2`] = `
"[33mâ€¼ [0mAssert application has exited
"
`;

exports[`AST renderers formatted text renderer should render test line results 3`] = `
"[34mÂ» [0mAssert application has exited
"
`;

exports[`AST renderers formatted text renderer should render test line results 4`] = `
"Â» Assert application has exited
"
`;

exports[`AST renderers formatted text renderer should render test line results 5`] = `
"[31mâœ– [0mAssert application has exited
"
`;

exports[`AST renderers plain text renderer should handle props 1`] = `
"prop name = expected value
prop name  
  > someJS();"
`;

exports[`AST renderers plain text renderer should handle props 2`] = `
"âœ– prop name prop name prop name prop name prop name ~ expected value expected value expected value expected value 
                                                      expected value                                              
                                                    â†’ expected value expected value expected value expected value 
                                                      expected value                                              
âœ– JavaScript expression                              
  > someJS(); someJS(); someJS(); someJS();        
  > someJS(); someJS(); someJS();                  
  > someJS(); someJS(); someJS();                  "
`;

exports[`AST renderers plain text renderer should render condition 1`] = `
"
condition: element My element exists
"
`;

exports[`AST renderers plain text renderer should render condition 2`] = `
"
condition: element My element exists
    âœ– prop name prop name prop name prop name prop name ~ expected value expected value expected value expected value 
                                                          expected value                                              
                                                        â†’ expected value expected value expected value expected value 
                                                          expected value                                              
    âœ– JavaScript expression                              
      > someJS(); someJS(); someJS(); someJS();        
      > someJS(); someJS(); someJS();                  
      > someJS(); someJS(); someJS();                  "
`;

exports[`AST renderers plain text renderer should render test line 1`] = `
"Assert element My element is visible
"
`;

exports[`AST renderers plain text renderer should render test line 2`] = `
"Assert element My element

  condition: element My element exists
"
`;

exports[`AST renderers plain text renderer should render test line results 1`] = `
"âœ– Assert application has exited

fail: Condition was not met"
`;

exports[`AST renderers plain text renderer should render test line results 2`] = `
"â€¼ Assert application has exited
"
`;

exports[`AST renderers plain text renderer should render test line results 3`] = `
"Â» Assert application has exited
"
`;

exports[`AST renderers plain text renderer should render test line results 4`] = `
"Â» Assert application has exited
"
`;

exports[`AST renderers plain text renderer should render test line results 5`] = `
"âœ– Assert application has exited
"
`;
